Support for building a Windows version with Microsoft Visual C++

JFLarvoire opened the pull request:
Hello,

This version builds ag for Windows with Microsoft Visual C++.  
It is loosely based on the version that KJK did a few years ago, and that is mentioned in your README.md.

Mine has the following changes compared to his:

- The ag sources, and the included libraries sources, are updated to the latest versions.  
   => Minimal changes to the ag src directory.
- The build procedure does not require any external development tool (besides Visual C++), and it's extremely simple.
- It should work with all versions of Visual C++ from VC++ 2010 to VC++ 2015. (VC++ 2017 is coming soon, and I'll add support asap.) (Older versions do not support C99, which is required by ag own sources.)
- It uses the MsvcLibX library to add the missing routines in Microsoft libraries, which brings many significant advantages compared to KJK's version:
- It uses UTF-8 internally, and outputs file paths correctly in any Windows code page.
  (This is very important to me, as my first name contains a non-ASCII character, and KJK's version did not display my HOME path correctly!)
- It fixes a number of bugs, including the 260-character limit in the basic Microsoft C libraries.
  (I do have systems with some paths longer than that!)
- Last but not least, it still builds successfully in Unix.

I know this looks like a huge merge, but the changes to the existing ag sources are actually very minor.
Most of the files are new files in the msc_libs subdirectory, and are copied from other projects.  
Note that all these other project have compatible licenses: Either Apache 2 like ag itself; Or BSD for some; Or LGPL for others.

Side note: I'm using PCRE regular expressions in UTF-8 mode, which fixes issue #1027 for Windows.  
I can work with you to fix it for Unix too.

There are explanations in the updated README.md, and more detailed ones in the msc_libs/README.md.

Before we go any further, maybe it'd be good to have the build process tested by other developers.
Is there a distribution list, where I post a message asking for volunteers?

Best regards,

Jean-François Larvoire

Commit:
Changes and new files to allow building ag.exe with Microsoft Visual C++.

Commit:
Updated instructions for building in Windows with Visual C++.

Commit:
Global ag configuration settings for building with MS Visual C++.

Commit:
Corrected a few markdown errors, and added details.

JFLarvoire committed:
Automatically detect and use the new common.mk if present.

JFLarvoire committed:
Added library files that git add forgot the first time.

JFLarvoire committed:
Added support for searching UTF-8 regular expressions.

JFLarvoire committed:
Added variable LOGDIR to control where to store the log.

JFLarvoire committed:
Leave the log file in the build directory itself.

JFLarvoire committed:
Added comments about recreating the character tables within Windows.

JFLarvoire committed:
Record both the console and system code pages.

JFLarvoire committed:
Rewrote fputcU() & fputsU() to write UTF16 to the console.

JFLarvoire committed:
Rewrote fwriteU() to write UTF16 to the console.

JFLarvoire committed:
Added fputs() series like for fputc.

JFLarvoire committed:
Fixed a wrong error code in case of failure.

JFLarvoire committed:
Removed a useless change.

JFLarvoire commented:
I tried running './format.sh reformat' to fix formatting, as suggested in the detailed reports.  
But this makes lots of changes to the existing files, in places that I've NOT changed.  
Just in case, I've checked in a minor change, removing a change that causes a warning in the report, and that I did not really need. (I did it while investigating the pthread 2.10 RC issue, and forgot to remove it.)  
What should I do now to fix these checks?

JFLarvoire committed:
Avoid trying to build ag.exe for DOS and WIN95, which are not supported yet.

JFLarvoire committed:
Added support for Visual Studio 2017.

JFLarvoire committed:
Get nmake.exe path from WIN32_PATH or WIN64_PATH.

JFLarvoire committed:
Convert Windows path separators to Unix path separators.

JFLarvoire committed:
For WIN32 builds, ignore path comparisons need to detect both Windows and Unix path separators, as both can be encountered.

JFLarvoire committed:
If WINVER is defined but empty, use SDK & linker defaults.

JFLarvoire committed:
Avoid defining environment STINCLUDE after exit.

JFLarvoire committed:
Added a comment.

JFLarvoire committed:
Don't automatically set the MSVCLIBX variable.

JFLarvoire committed:
Changed the iWideConsoleInitialized array elements type to int.

JFLarvoire committed:
Added the output of Windows/Visual C++ builds.

JFLarvoire committed:
All MsvcLibX files are now included.

JFLarvoire committed:
Fix build if environment variable OS is not defined.

JFLarvoire committed:
Changed text output to stdout or stderr to the console, to be UTF-16 Unicode.

jschpp committed:
clang-format

jschpp commented:
I added a PR to your repo with clang-format 3.8 JFLarvoire/the_silver_searcher#2

JFLarvoire committed:
Merge pull request #2 from jschpp/pr/1064

clang-format

JFLarvoire committed:
The -D option also enables MsvcLibX debugging in the DEBUG build.

JFLarvoire committed:
Merge branch 'master' of git+ssh://github.com/JFLarvoire/the_silver_searcher

jschpp commented:
Nicely done. And the tests run without errors 👍 
Although there are still some compiler warnings

JFLarvoire committed:
Restored the initial formatting, but with tabulations changed to spaces.

JFLarvoire committed:
Moved unlink() & rmdir() definitions from sys/stat.h to unistd.h.

JFLarvoire committed:
Moved unlink*() & rmdir*() from lstat.c to their own source files.

JFLarvoire committed:
Added sources unlink.c and rmdir.c.

JFLarvoire committed:
Variable OS defines which OS versions need to be built.

JFLarvoire committed:
Avoid warnings.

JFLarvoire committed:
Added missing dependencies.

JFLarvoire committed:
Bug fix: Only change Xlation when writing to a valid file.

JFLarvoire committed:
Bug fix: Only change Xlation when writing to a valid file.

JFLarvoire committed:
Include stdio.h, to get the UTF-8 version of printf.

JFLarvoire committed:
Bug fix: _setmodeX() now checks its input values validity.
Bug fix: Static variables in fputcM must be thread-local.

JFLarvoire committed:
Bug fix in ResolveLinks: First convert relative paths to absolute paths.

JFLarvoire committed:
Updated comments.

JFLarvoire committed:
Avoid getting Windows version twice to be thread safe.

JFLarvoire committed:
Removed a stray constant. (Why wasn't it caught by the compiler?!)

JFLarvoire committed:
No need to fully resolve all links in the pathname. Resolving the tail links is sufficient, and much faster.
And only resolve links if it's a link.

JFLarvoire committed:
Rewrote DEBUG_PRINTF() and similar macros to generate a string, and output it in a single call to puts().
This is useful for multi-threaded programs, that need to use a semaphore for synchronizing debug output from multiple threads.
Added macro SET_DEBUG_PUTS().

JFLarvoire committed:
Added ResolveTailLinks*() prototypes.

JFLarvoire committed:
Added routines TrimTailSlashesW() and ResolveTailLinks*().

JFLarvoire committed:
Include stdio.h, to get the UTF-8 version of printf.

JFLarvoire committed:
MsvcLibX debug output now goes though ag_debug_puts -> log_debug

JFLarvoire committed:
Define __thread for Microsoft compilers.

JFLarvoire committed:
Change worker_id to a global thread variable. It's now used in log_debug, to allow filtering the debug output by thread.

JFLarvoire committed:
log_debug displays the worker threads IDs. This allows filtering the debug output by thread.

JFLarvoire committed:
Moved the __thread definition for Microsoft compilers to util.h.

JFLarvoire committed:
The build has been tested in Visual C++ 2017.

JFLarvoire committed:
Reformat sources to satisfy clang-format ./format.sh test.

JFLarvoire commented:
The problem is not with my code anymore, it's with the checking process:
`./format.sh reformat` simply breaks the Windows batch file and the two make files that I've hade to add into the src directory.
How can I fix format.sh so that it checks only *.c and *.h files?

JFLarvoire commented:
Unrelated: I've had a nasty issue with threading in MsvcLibX, which forced me to make a change in ag's log_debug routine: 
log_bug now displays the worker thread ID ahead of the debug string. This allows filtering the debug output by thread number, to make sense of who's doing what.
I've left this in, as the change is relatively unobtrusive, and I feel this might be useful to others.
The issue in MsvcLibX is now fixed :-) (Some code used a static variable, which was not thread safe.)

JFLarvoire committed:
Do not attempt reformating src/*.bat nor src/*.mak, as these are not C source files

JFLarvoire commented:
I finally modified format.sh to only reformat src/*.c and src/*.h files, and the test now passes :-)

JFLarvoire committed:
Revert to the old implementation for DOS, as the big DEBUG_GLOBALS macro sometimes chokes the DOS compiler.

JFLarvoire committed:
Added a comment about the Windows Glyph List.

JFLarvoire committed:
Added the new IO_REPARSE_TAG_LXSS_SYMLINK tag.

JFLarvoire committed:
Added puts(), and ANSI versions of fputc, fputs, fwrite.

JFLarvoire committed:
Added puts().

JFLarvoire committed:
Bug fix: realloc() sometimes moves the _acmdln buffer.

JFLarvoire committed:
Create a link from src\bin dir to the silver searcher's main bin dir.

JFLarvoire committed:
The ag.exe output directory now is also bin\win32 and bin\win64.

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

# Conflicts:
#	src/util.h
Fixed by adding an additional conditional compilation case for MSVCLIBX.

JFLarvoire committed:
Added prototypes for vfprintfM(), fprintfM() and printfM().

JFLarvoire committed:
Added routines fprintfM() and printfM().

JFLarvoire committed:
Added support for two text file encodings.

JFLarvoire committed:
Added a description of the ag.exe program features.
Added a description of the process for building debug versions of ag.exe.

JFLarvoire committed:
Updated print_version() to display the OS version, debug version, and supported encodings.

JFLarvoire committed:
Changed the package version to 2.0.0.

JFLarvoire committed:
Reformatted the source to satisfy clang-format.

JFLarvoire committed:
Fixed two wrong data types, which surprisingly did not prevent the Windows version from working.

JFLarvoire committed:
Satisfy clang-format

JFLarvoire committed:
Satisfy clang-format

JFLarvoire committed:
Satisfy clang-format

JFLarvoire committed:
Satisfy clang-format

JFLarvoire committed:
Satisfy clang-format

JFLarvoire committed:
Yet another try at running clang-format.

JFLarvoire committed:
Yet another run of clang-format.

JFLarvoire commented:
I'll need help setting up clang-format:

I've uploaded a couple of days ago a set of sources that work... But "make test" fails on github because of clang-format errors.
I've struggled the whole day yesterday trying to make it work on my PC, but no matter what I do, this fails:

- In Windows subsystem for Linux, which is a plain Ubuntu, the ag build works,
  but clang-format segfaults after modifying a couple of files.
- I tried installing clang-format in a CentOS VM, to rebuild everything in a true Linux environment...
  Again, clang-format segfaults after modifying files differently!
- Finally I realized that a Windows version of clang-format does exist. I installed it.
  This one successfully processes all files (Bypassing format.sh, and running clang-format directly with the same options), but with still different results!

And in ALL 3 CASES, when I push the modified files I have to github, the version of clang-format there rejects them as wrongly formatted!!!

Could anybody explain me how to setup a development environment with clang-format, that works like the one used by the github build system?

Jean-François

osimola commented:
I had the same problem with clang-format. Different versions just format differently and there is no knowing which one Travis is running.

I solved the problem by copying the suggested diff from error report and applying it to my copy.

JFLarvoire committed:
Satisfy clang-format

JFLarvoire commented:
Thanks @osimola, all checks have passed at last :-)

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

JFLarvoire committed:
Merged updates from MsvcLibX.

JFLarvoire committed:
Updated the version to 2.1.0

JFLarvoire committed:
Added a short term workaround for MsvcLibX lack of support for Unicode on stdin.

JFLarvoire committed:
An NMake file for making ag libs, ag, and its releases.

JFLarvoire committed:
Satisfy clang_format.

JFLarvoire committed:
Fix for bug #1158

JFLarvoire committed:
Merge branch 'stats_64bits_size' of git+ssh://github.com/JFLarvoire/the_silver_searcher

JFLarvoire committed:
Updated comments and log messages.

JFLarvoire committed:
Added option -? to get help.

JFLarvoire committed:
Updated the version to 2.1.1.

JFLarvoire committed:
Merged in changes for long pathnames (>=260 characters) support.

JFLarvoire committed:
Define the ag.exe version in a single place.

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

JFLarvoire committed:
Updated to the 2017-10 pthreads 2.10 RC, with a few patches.

JFLarvoire committed:
Updated to the 2017-10-31 version.

JFLarvoire committed:
Changed OUTDIR default to the bin subdirectory.

JFLarvoire committed:
Added macro DEBUG_QUIET_LEAVE().

JFLarvoire committed:
Changed the output file name to $(PROGRAM).objects. Fixes the bug which causes libraries to be rebuilt even if no source changed.

JFLarvoire committed:
Added inference rules to build a DLL.

JFLarvoire committed:
Added rules to build a DLL.
Renamed *.LNK as *.link, and *.LST as *.list.
Added an optional $(POST_LINK_CMD).

JFLarvoire committed:
Bug fix: Avoid breaking the cmd windows title.
Added a search in the PATH, to be forward-compatible.
Added the :CheckDir subroutine to search consistently.

JFLarvoire committed:
The pthreads.lib library is now called pthreadsVC2.lib.

JFLarvoire committed:
Define MD_OUTDIR to trigger the creation of a bin\ link instead of a directory.
Even if the right tools are available, avoid trying to build ag.exe for DOS and WIN95, which are not supported yet.
The pthreads.lib output path has changed.

JFLarvoire committed:
The bin link creation is handled by defining MD_OUTDIR in a configure.msc_libs.bat file.

JFLarvoire committed:
Added a special build rule for the pthread library.

JFLarvoire committed:
Corrected an error in the debug build instructions.

JFLarvoire committed:
Changed line endings from LF to CRLF.

JFLarvoire committed:
Removed the blank line that I had to add to allow the previous commit.

JFLarvoire committed:
File name case changed.

JFLarvoire committed:
Realigned some macro definitions, which were aligned differently from SysToolsLib original!?! No functional code change.

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

JFLarvoire committed:
Renamed the "msc_libs" directory back to the standard "win32" name.

JFLarvoire committed:
Added debug macros RETURN_PTR(p), RETURN_LONG(l), RETURN_CSTRING(s), RETURN_CPTR(p).

JFLarvoire committed:
Also search in win32 subdirectories, which are standard in many Unix+Windows open source projects.
Fixed several :# comments, which should have been &:#.
Improved comments and the debugging output.

JFLarvoire committed:
Removed set "OUTDIR=bin" as this is now the default.

JFLarvoire committed:
Updated comments

JFLarvoire committed:
Changed the output make file name.

JFLarvoire committed:
Added $(LIBRARIES) to the $(PROGRAM).* dependency list in make files, so that $(PROGRAM).* is relinked if one of its libraries changes.
The LIBRARIES variable is generated by lib2libs.mak/lib2libs.bat based on the LIB and LIBS variables.

JFLarvoire committed:
Added a message when creating config.h.

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

# Conflicts:
#	src/main.c
#	src/util.h

JFLarvoire committed:
MsvcLibX and make system updates as of 2018-03-11

JFLarvoire committed:
Changed the pthreads.lib output directory

JFLarvoire committed:
Changed the pthreads.lib name

JFLarvoire committed:
Optionally build libs as one source too.

JFLarvoire committed:
Switch back to using one source. The saving was minimal for ag.exe, and there were make command-line size issues on some systems.

JFLarvoire committed:
Bug fix for static builds: Added memory heap initializations if needed.

JFLarvoire committed:
Use %STINCLUDE% instead of $(STINCLUDE), to make the command line shorter.

JFLarvoire committed:
Use junction.exe, if available, instead of mklink.exe /j.

JFLarvoire committed:
Fixed the last change for junction.exe support.

JFLarvoire committed:
Added a source-release target.

JFLarvoire committed:
Fix the search for the Windows Kit 8.0.

JFLarvoire committed:
Use /MD for compilations for DLLs, and /MT for static libraries.

JFLarvoire committed:
Added an introduction describing the specificities of this fork.

JFLarvoire committed:
Added a bullet item about the make improvements.

JFLarvoire committed:
Display all libraries versions.

JFLarvoire committed:
Added several missing DEBUG_xxx_COMMENT() macros.
Added macro DEBUG_WPRINTF().
Added macros DEBUG_WENTER() and DEBUG_WLEAVE().

JFLarvoire committed:
Updated to MsvcLibX 2018-05-02.

JFLarvoire committed:
Updated to PCRE 8.42 2018-03-20.

JFLarvoire committed:
Document the latest versions of pcre and pthreads4w libraries used.

JFLarvoire committed:
Moved the display of the text file encodings supported from --version to --help.
Updated --version to only display if the two-encodings feature is supported or not.
Updated --help to list contributors to this Windows version, and where it is available.

JFLarvoire committed:
Fixes #4 - Arguments to the fprintfM() function were ordered incorrectly, and this caused the --before and --context arguments to crash.

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

JFLarvoire committed:
Updated version to 2.2.0.

JFLarvoire committed:
Merged in the latest MsvcLibX version.

JFLarvoire committed:
Merged in pthread4w 3.0.0 changes.

JFLarvoire committed:
Bug fix: Lines were output with two CR then one LF.

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

JFLarvoire committed:
Merged in the latest SysToolsLib and MsvcLibX changes.

JFLarvoire committed:
Convert path separators in error messages back to \ for Windows.

JFLarvoire committed:
Added MsvcLibX/Windows port version and date.

JFLarvoire committed:
Expand wildcards in command-line arguments for Windows.
Display a more detailed version for the Windows port with option -V.

JFLarvoire committed:
Document the latest features.

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

JFLarvoire committed:
Merged in MsvcLibX updates

JFLarvoire committed:
Added the 7-Zip LZMA SDK to the list of known SDKs.

JFLarvoire committed:
'make release' now generates an ag_setup.exe and a winget manifest.

JFLarvoire committed:
Document the ability to install using winget.

JFLarvoire committed:
Describe the winget package management for the Silver Searcher

JFLarvoire committed:
Merged in changes from SysToolsLib C\include.

JFLarvoire committed:
Upgraded the PCRE library to version 8.44.

JFLarvoire committed:
Updated the project version and date.

JFLarvoire committed:
Pthreads and zlib are still up-to-date as of June 2020.

JFLarvoire committed:
Improved the winget documentation.

JFLarvoire committed:
Added support for versions with 4 parts: X.Y.Z.T

JFLarvoire committed:
Merge remote-tracking branch 'ggreer_version/master'

JFLarvoire committed:
Prevent MsvcLibX macro redefinition warnings in zlib

JFLarvoire committed:
MsvcLibX and make system updates as of 2021-05-15

JFLarvoire committed:
Convert \uXXXX & \UXXXXXXXX escape sequences in the search pattern, except when using option -Q/--literal.
Added help about this, and about the output encoding rules.

JFLarvoire committed:
Removed leftover debug code

JFLarvoire committed:
Added a warning to prevent people from editing config.h.

JFLarvoire committed:
Added an extra lpUsedDefault argument to routine ConvertString, to know if there were characters impossible to convert.

JFLarvoire committed:
Added option --verbose to display the pattern string generated after \uXXXX replacement;
and fixed a bug when that escape sequence was last in the pattern.

JFLarvoire committed:
Make sure that if some Unicode characters cannot be converted to the CP_ACP code page, then the regular expression cannot match anything in CP_ACP input text.

JFLarvoire committed:
ag.exe updated today

JFLarvoire committed:
Avoid warnings due to HAVE_STRUCT_TIMESPEC redefinition

JFLarvoire committed:
Updated MsvcLibX to version 2021-06-03

JFLarvoire committed:
Added support for UTF-16 files, and the dynamic detection of the encoding on stdin

JFLarvoire committed:
New features in the 2.2.4 release

JFLarvoire committed:
Restores the initial console output color when aborting the program with Ctrl-C

JFLarvoire committed:
Another change today

JFLarvoire committed:
Register The Silver Searcher installation, so that winget can list it and uninstall it.

JFLarvoire committed:
Removed leftover temporary debug directives

JFLarvoire committed:
Detect Ag installations done by Chocolatey and Scoop, and abort this installation if one was found.
Fixed the administrator detection in 32-bits shells in 64-bits Windows. (Which is the case within ag_setup.exe.)
Pause before exiting after an error, to give time to read the error message from within ag_setup.exe.

JFLarvoire committed:
Make release now also generates architecture-specific zip files, for use by Chocolatey, etc.

JFLarvoire committed:
Release 2.2.5

JFLarvoire committed:
Add new features in setup.bat

JFLarvoire committed:
Updated manifest syntax; Added the x64 architecture entry.

JFLarvoire committed:
Updated winget and Chocolatey packages creation instructions.

JFLarvoire committed:
Fixed the hash generation for the latest winget version

JFLarvoire committed:
Added option -y

JFLarvoire committed:
Added information about how to debug Ag setup.

JFLarvoire committed:
Fixed the registry update from within ag_setup.exe

JFLarvoire committed:
Actually no, must not add an x64 architecture entry, as this does not pass the winget validation.

JFLarvoire committed:
Improved test and debug instructions

JFLarvoire committed:
Reordered other package managers detection, so that comments make sense again

JFLarvoire committed:
Updated list of files to ignore.

JFLarvoire committed:
Merged in SysToolsLib version 1.23 2026-02-21

JFLarvoire committed:
Fixed warnings detected by Visual C++ 2026.

JFLarvoire committed:
Define dictionary management routines here, as they're not part of MsvcLibX anymore.

JFLarvoire committed:
Updated the PACKAGE_VERSION to 2.2.6.

JFLarvoire committed:
Updated comments

JFLarvoire committed:
Adapted the STINCLUDE detection for Ag 2.2.6 without sub-modules like NMaker.

JFLarvoire committed:
Removed configure.bat and make.bat proxy copies from the git database, and added them to the .gitignore file.

JFLarvoire committed:
Moved CLEAN_DIRS execution before the $(OUTDIR) cleanup, to increase the chances of being able to remove $(OUTDIR).
Make clean now displays the relative paths it cleans.
Added the CLEAN_OUTDIRS optional variable.

JFLarvoire committed:
Adapted the STINCLUDE detection for projects that still include NMaker and MsvcLibX copies, & not as git modules.

JFLarvoire committed:
Define the mostlyclean and distclean targets.

JFLarvoire committed:
In the $(CONFIG_H): rule, handle the case of projects that have their own config.h file.
Do not define _DLL in the .mak.lib inference rule.
Define the mostlyclean and distclean targets.

JFLarvoire committed:
Added a CLEAN_OUTDIRS directive, to make sure that win32/pthreads/bin is deleted by `make clean`.

JFLarvoire committed:
Updated build instructions.

JFLarvoire committed:
Updated the 2.2.6 release date.

JFLarvoire committed:
Use NMaker as a git submodule

JFLarvoire committed:
Updated PCRE 1 from 8.44 to 8.45.

JFLarvoire committed:
Ignore diff & patch files.

JFLarvoire committed:
Updated zlib from version 1.2.11 to version 1.3.2.

JFLarvoire committed:
Fixed the FILE_ID_128 conditional redefinition.

JFLarvoire committed:
Ignore old library versions kept for reference.

JFLarvoire committed:
Get the latest NMaker also searching for NMINCLUDE in [...]\WIN32\NMaker\include.

JFLarvoire committed:
Updated sourceforge.net pthreads4w 3.0.0 to GitHub GerHobbelt/pthread-win32 version 3.0.3.1.

JFLarvoire committed:
Correct the MSVCLIBX dependency.

JFLarvoire committed:
No need for a special case for pthread anymore.

JFLarvoire committed:
Updated information about pcre and pthreads.

JFLarvoire committed:
Avoid two duplicate targets warnings for win32 & src.

JFLarvoire committed:
Added the --unix option for testing ag.exe in WSL, with output looking like that of the Unix version of ag.

JFLarvoire committed:
Oops, correct the last commit adding the --unix option.

JFLarvoire committed:
Don't convert / to \ in the debugging output, because we want to know exactly what's executed.

JFLarvoire committed:
Fix for bug #7 ".gitignore patterns respected by Linux ag don't work in Windows ag"

JFLarvoire committed:
Version 2.2.7

JFLarvoire committed:
Information about the cram test system, and how to test The Silver Searcher using it.

JFLarvoire committed:
Corrected the last commit #fdcd47f03c6fba8311a626006c2dc79ed606d523 "Don't convert / to \ in the debugging output, because we want to know exactly what's executed"

JFLarvoire committed:
Make release now adds the NEWS file to every zip file.
Make release now adds MsgBox.exe, if present, to ag_setup.exe.

JFLarvoire committed:
Release history

JFLarvoire committed:
Added support for ARM and ARM64 systems.
Added a final message box summarizing the result.

JFLarvoire committed:
Removed the dependency on wmic.exe, which is not always installed anymore.
Correctly detect the ARM builds.
Updated the final message.

JFLarvoire committed:
Releasing version 2.2.7 today.

JFLarvoire committed:
Updated release news for version 2.2.7.
